internal/provider/license_resource.go

```diff
@@ -0,0 +1,194 @@
+				Computed:            true,
+				PlanModifiers: []planmodifier.Int32{
+					int32planmodifier.UseStateForUnknown(),
+				},
+			},
+			"expires_at": schema.Int64Attribute{
+				MarkdownDescription: "Unix timestamp of when the license expires.",
+				Computed:            true,
+			},
+			"license": schema.StringAttribute{
+				MarkdownDescription: "A license key for Coder.",
+				Required:            true,
+				Sensitive:           true,
+				PlanModifiers: []planmodifier.String{
+					stringplanmodifier.RequiresReplace(),
+				},
+			},
+		},
+	}
+}
+
+func (r *LicenseResource) Configure(ctx context.Context, req resource.ConfigureRequest, resp *resource.ConfigureResponse) {
+	// Prevent panic if the provider has not been configured.
+	if req.ProviderData == nil {
+		return
+	}
+
+	data, ok := req.ProviderData.(*CoderdProviderData)
+
+	if !ok {
+		resp.Diagnostics.AddError(
+			"Unexpected Resource Configure Type",
+			fmt.Sprintf("Expected *CoderdProviderData, got: %T. Please report this issue to the provider developers.", req.ProviderData),
+		)
+
+		return
+	}
+
+	r.data = data
+}
+
+func (r *LicenseResource) Create(ctx context.Context, req resource.CreateRequest, resp *resource.CreateResponse) {
+	var data LicenseResourceModel
+
+	// Read Terraform plan data into the model
+	resp.Diagnostics.Append(req.Plan.Get(ctx, &data)...)
+
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	client := r.data.Client
+
+	license, err := client.AddLicense(ctx, codersdk.AddLicenseRequest{
+		License: data.License.ValueString(),
+	})
+	if err != nil {
+		resp.Diagnostics.AddError("Client Error", fmt.Sprintf("Unable to add license, got error: %s", err))
+		return
+	}
+	data.ID = types.Int32Value(license.ID)
+	expiresAt, err := license.ExpiresAt()
+	if err != nil {
+		resp.Diagnostics.AddError("Client Error", fmt.Sprintf("Unable to parse license expiration, got error: %s", err))
+		return
+	}
+	data.ExpiresAt = types.Int64Value(expiresAt.Unix())
+
+	// Save data into Terraform state
+	resp.Diagnostics.Append(resp.State.Set(ctx, &data)...)
+}
+
+func (r *LicenseResource) Read(ctx context.Context, req resource.ReadRequest, resp *resource.ReadResponse) {
+	var data LicenseResourceModel
+
+	// Read Terraform prior state data into the model
+	resp.Diagnostics.Append(req.State.Get(ctx, &data)...)
+
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	licenses, err := r.data.Client.Licenses(ctx)
+	if err != nil {
+		resp.Diagnostics.AddError("Client Error", fmt.Sprintf("Unable to list licenses, got error: %s", err))
+		return
+	}
+
+	found := false
+	for _, license := range licenses {
+		if license.ID == data.ID.ValueInt32() {
+			found = true
+			expiresAt, err := license.ExpiresAt()
+			if err != nil {
+				resp.Diagnostics.AddError("Client Error", fmt.Sprintf("Unable to parse license expiration, got error: %s", err))
+				return
+			}
+			if expiresAt.Before(time.Now()) {
+				resp.Diagnostics.AddError("Client Error", "License has expired.")
+				return
+			}
```

Comment:
I think this check should be removed, we don't check expiry on the server except for computing entitlements. You can insert and view expired licenses AFAIK